lib/browser-specific.js

```diff
@@ -0,0 +1,247 @@
+'use strict';
+
+/**
+ * Implements functionality to report on how many WPT tests fail only on one
+ * browser (aka 'browser-specific failures).
+ */
+
+// A helper class providing an iterator-like interface to either an Object with
+// enumerable properties, or an Array. Iterates in a sorted order; alphabetical
+// in the case of an Object, or in the order defined by |comparatorFunc| in the
+// case of an Array.
+class IteratorHelper {
+  constructor(arrOrObject, comparatorFunc = null) {
+    this.currentIndex = 0;
+    this.values = arrOrObject;
+
+    if (this.values instanceof Array) {
+      this.keys = null;
+      this.maxIndex = this.values.length;
+      this.values.sort(comparatorFunc);
+    } else {
+      this.keys = Object.keys(this.values);
+      this.maxIndex = this.keys.length;
+      this.keys.sort();
+    }
+  }
+
+  hasNext() {
+    return this.currentIndex < this.maxIndex;
+  }
+
+  next() {
+    if (!this.hasNext()) {
+      throw new Error('Cannot move next; out of values');
+    }
+    this.currentIndex++;
+  }
+
+  currentKey() {
+    if (this.keys === null) {
+      throw new Error('Cannot get key of an Array iteration');
+    }
+    return this.keys[this.currentIndex];
+  }
+
+  currentValue() {
+    if (this.keys === null) {
+      return this.values[this.currentIndex];
+    }
+    return this.values[this.currentKey()];
+  }
+}
+
+// Scores a particular WPT test for a set of browsers, updating |scores|. It is
+// assumed that the order of |browserTests| and |scores| is the same.
+function scoreTest(browserTests, scores) {
+  // Some WPT tests contain multiple 'subtests' (e.g. most jsharness tests),
+  // whilst others are just a single conceptual test (e.g. reftests).
+  //
+  // Tests without subtests are scored as a simple 0-or-1 for each failing
+  // browser (0 if any other browser also fails, 1 if no other browser fails).
+  // When there are subtests, we do a similar calculation per-subtest, but
+  // normalize the results by the number of subtests in the test. This stops
+  // tests with thousands of subtests from dominating the results.
+  if (browserTests.every(tests => !tests.subtests || tests.subtests.length == 0)) {
+    let failed = [];
+    for (let i = 0; i < browserTests.length; i++) {
+      // TODO: Determine exact criteria for a BSF at a test level.
```

Comment:
Ok, so let's talk this through.

## Statuses

At the test level, there are the following statuses we could see (from [mozlog](https://searchfox.org/mozilla-central/rev/c61720a7d0c094d772059f9d6a7844eb7619f107/testing/mozbase/mozlog/mozlog/logtypes.py#192-205)):

- `PASS` - test was a reftest or single page test, and passed
- `FAIL` - test was a reftest or single page test, and failed
- `OK` - test should had subtests, had no top-level errors
- `ERROR` - test had a top-level error. It **may** have subtests, which **may** have any status (as far as I can tell, based on searching wpt.fyi. Not validated against wpt-results data).
- `TIMEOUT` - test timed out. It may or may not have subtests. If it does, some subtests may have other statuses but one subtest **should** be TIMEOUT (I think)
- `CRASH` - the browser crashed. This test should **not** have subtests.
- `ASSERT` - unsure if we ever have this (not verified in wpt-results data).
- `PRECONDITION_FAILED` - test failed a precondition. It may or may not have subtests **(I think?)**; if it does, they may have any status.
- `NOTRUN` - doesn't show up for tests as far as I know
- `SKIP` - test marked as should be skipped (rare to non-existant?)

At the subtest level, there are the following statuses we could see:

- `PASS` - subtest passed
- `FAIL` - subtest failed
- `OK` - not seen for subtests
- `ERROR` - not seen for subtests
- `TIMEOUT` - subtest timed out
- `CRASH` - not seen for subtests
- `ASSERT` - unsure if we ever have this (not verified in wpt-results data).
- `PRECONDITION_FAILED` - subtest failed a precondition
- `NOTRUN` - subtest was registered, but never run (always because the test ended up with TIMEOUT?)
- `SKIP` - subtest marked as should be skipped (rare to non-existant?)

The `MISSING` status that one sees on wpt.fyi isn't a real status, it is just what we put when we have no results for that subtest from that product (but do have results from other products). So in wpt-results/, it just doesn't exist.

## Tests without subtests

Let's start with the case where no test has any subtests listed. I would say this is either definitely a reftest/single-page test, or its a completely broken subtest-test where all browsers have ERROR status. Let's ignore the latter (not a BSF), and consider the former. 

PASS and FAIL seem uncontroversial. I feel like ERROR, TIMEOUT, and CRASH should all count as FAIL... there's an argument that they can be caused by test problems, but if a test is causing a BSF then we still want to uncover why (and in theory it shouldn't be systematic, unless we have systematic test issues). We should assert that ASSERT and NOTRUN are never seen. Finally, I think PRECONDITION_FAILED and SKIP should be considered 'neutral', such that they either neither counted as a BSF, nor can they cause one for another browser. That is, neither of the below are BSFs:

[PASS, FAIL, PRECONDITION_FAILED]  # not a BSF for browser #2
[PASS, PASS, PRECONDITION_FAILED]  # not a BSF for browser #3

## Tests with subtests

Now we're in the situation where some browser has subtests for the given test. We've chosen to take the intersection of available subtests as the 'least wrong' metric. There are some problems with that (if a browser has a bug so is TIMEOUT with a bunch of missing subtests, it isn't penalized for that), but it avoids other problems so *shrug*. (Yes, I should be able to defend what problems it does avoid, which I can't bring to mind right now...).

So for each subtest that appears in all browsers, what should we count statuses as? Again, PASS/FAIL seem uncontroversial. We should assert that OK, ERROR, CRASH, and ASSERT are not seen. I think we should map TIMEOUT to FAIL, and PRECONDITION_FAILED, NOTRUN, and SKIP to the same neutral concept considered above.



Thoughts on all of the above welcome.

Reply:
I agree completely regarding "tests without subtests".

For tests with subtests, my intuition is to take the union of all subtests if we are sure that we are comparing aligned runs. What'd be the problem of that? Or what advantage does intersection have over union? And for `NOTRUN`, I think we probably need to consider it as a failure. It's often the result of timeout. If we don't count them as failures, then we might have just a single timeout subtest and all the others are pass/neutral, which seems too lenient?

Reply:
Philip and I debated the union vs intersection question a while ago. From my notes back then:

> **Stephen**: Union of subtests, and then score by number of browser-specific failures with 	number of subtests as the denominator. And PASS PASS not-PASS, not other statuses.
> **Philip**: If you imagine this as 3-slices from the ⅓ of the interop chart, does it add up properly? Problem with union is when it’s permissible to have different behavior per spec, could make it impossible to score 100% on the metric. 
> - pass/pass/missing for safari shows up 405 tests, ~9k subtests. 2.5k are webaudio, 1k are web-animations.
> - We could do pass/pass/{!missing, !pass} ?
>
> Discussion, agreed on pass/pass/{!missing, !pass} (which is intersection)

So at least one concern at the time was tests that generate their subtests based on permissibly different behavior, which could lead to penalizing browsers incorrectly. @foolip may have a better argument here too

Reply:
On `PRECONDITION_FAILED`, a test-level `PRECONDITION_FAILED` should be possible together with subtests with any other failure, if the precondition is failed at the "right" time. Looks like I didn't add any tests for it in infrastructure/expected-fail/, however.

Reply:
I agree on the proposed pass/fail/neutral categorization of statuses in "tests without subtests", where any neutral status (like `PRECONDITION_FAILED`) means that test can't be a BSF.

For subtests, an explicitly neutral statuses like `PRECONDITION_FAILED` also shouldn't result in a BSF, and I think the tricky case is `NOTRUN` and entirely missing results. The options are to treat them as neutral or as failing.

Treating them as neutral taking the intersection of subtests, and means the inconsistently-reported subtests will never be BSF. What remains will be more actionable, but it will likely also make the metric a bit less stable, where fixing one subtest can cause a bunch of new subtests to appear and take part in the scoring. That could create the wrong incentive in the local case.

Treating them as failing is roughly the union of subtests, but two other browsers still need have run and passed the subtest for it to count as a BSF. For the browser that didn't run it, but there will still be some bogus BSFs. The upside is fixing problem will be an improvement in the metric, which is a good property to have.

Based on these theoretical properties and hunches I think I'd favor the `PASS` / `PASS` / not-`PASS`-or-`PRECONDITION_FAILED` treatment for subtests, roughly the union.

To inform this a bit more, can we look at what the graphs over time for both methods would look like? Where there are differences, maybe one approach aligns better with what we think should be valued?

Reply:
(Fun side-note; I was working on adding neutral values and testing out union, and discovered that we do have subtests with an ERROR status. The WebDriver python tests appear to be able to generate ERROR statuses for subtests, e.g. see https://wpt.fyi/results/webdriver/tests/permissions/set.py?label=master&label=experimental&aligned)